subsys/esb/esb.c

```diff
@@ -547,25 +482,37 @@ static void sys_timer_init(void)
 
 static void ppi_init(void)
 {
-	NRF_PPI->CH[CONFIG_ESB_PPI_TIMER_START].EEP =
-		(uint32_t)&NRF_RADIO->EVENTS_READY;
-	NRF_PPI->CH[CONFIG_ESB_PPI_TIMER_START].TEP =
-		(uint32_t)&ESB_SYS_TIMER->TASKS_START;
-
-	NRF_PPI->CH[CONFIG_ESB_PPI_TIMER_STOP].EEP =
-		(uint32_t)&NRF_RADIO->EVENTS_ADDRESS;
-	NRF_PPI->CH[CONFIG_ESB_PPI_TIMER_STOP].TEP =
-		(uint32_t)&ESB_SYS_TIMER->TASKS_SHUTDOWN;
-
-	NRF_PPI->CH[CONFIG_ESB_PPI_RX_TIMEOUT].EEP =
-		(uint32_t)&ESB_SYS_TIMER->EVENTS_COMPARE[0];
-	NRF_PPI->CH[CONFIG_ESB_PPI_RX_TIMEOUT].TEP =
-		(uint32_t)&NRF_RADIO->TASKS_DISABLE;
-
-	NRF_PPI->CH[CONFIG_ESB_PPI_TX_START].EEP =
-		(uint32_t)&ESB_SYS_TIMER->EVENTS_COMPARE[1];
-	NRF_PPI->CH[CONFIG_ESB_PPI_TX_START].TEP =
-		(uint32_t)&NRF_RADIO->TASKS_TXEN;
+#ifdef DPPI_PRESENT
+	nrfx_dppi_channel_alloc(&ppi_ch_radio_ready_timer_start);
+	nrfx_dppi_channel_alloc(&ppi_ch_radio_address_timer_stop);
+	nrfx_dppi_channel_alloc(&ppi_ch_timer_compare0_radio_disable);
+	nrfx_dppi_channel_alloc(&ppi_ch_timer_compare1_radio_txen);
+
+	NRF_RADIO->PUBLISH_READY          = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | ppi_ch_radio_ready_timer_start;
+	ESB_SYS_TIMER->SUBSCRIBE_START    = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | ppi_ch_radio_ready_timer_start;
+	NRF_RADIO->PUBLISH_ADDRESS        = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | ppi_ch_radio_address_timer_stop;
+	ESB_SYS_TIMER->SUBSCRIBE_SHUTDOWN = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | ppi_ch_radio_address_timer_stop;
+	ESB_SYS_TIMER->PUBLISH_COMPARE[0] = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | ppi_ch_timer_compare0_radio_disable;
+	NRF_RADIO->SUBSCRIBE_DISABLE      = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | ppi_ch_timer_compare0_radio_disable;
+	ESB_SYS_TIMER->PUBLISH_COMPARE[1] = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | ppi_ch_timer_compare1_radio_txen;
+	NRF_RADIO->SUBSCRIBE_TXEN         = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | ppi_ch_timer_compare1_radio_txen;
+#else
+	nrfx_ppi_channel_alloc(&ppi_ch_radio_ready_timer_start);
+	nrfx_ppi_channel_alloc(&ppi_ch_radio_address_timer_stop);
+	nrfx_ppi_channel_alloc(&ppi_ch_timer_compare0_radio_disable);
+	nrfx_ppi_channel_alloc(&ppi_ch_timer_compare1_radio_txen);
+
+	nrfx_ppi_channel_assign(ppi_ch_radio_ready_timer_start, 
+		(uint32_t)&NRF_RADIO->EVENTS_READY, (uint32_t)&ESB_SYS_TIMER->TASKS_START);
+	nrfx_ppi_channel_assign(ppi_ch_radio_address_timer_stop, 
+		(uint32_t)&NRF_RADIO->EVENTS_ADDRESS, (uint32_t)&ESB_SYS_TIMER->TASKS_SHUTDOWN);
+	nrfx_ppi_channel_assign(ppi_ch_timer_compare0_radio_disable, 
+		(uint32_t)&ESB_SYS_TIMER->EVENTS_COMPARE[0], (uint32_t)&NRF_RADIO->TASKS_DISABLE);
+	nrfx_ppi_channel_assign(ppi_ch_timer_compare1_radio_txen, 
+		(uint32_t)&ESB_SYS_TIMER->EVENTS_COMPARE[1], (uint32_t)&NRF_RADIO->TASKS_TXEN);
```

Comment:
Have you considered using nrfx abstraction over PPI / DPPI?

https://github.com/nrfconnect/sdk-hal_nordic/blob/master/nrfx/helpers/nrfx_gppi.h

If API is flexible enough to cover ESB logic, then we will have uniform code for both nRF52 and nRF53. You can see how it is used in the DTM example:

https://github.com/nrfconnect/sdk-nrf/blob/master/samples/bluetooth/direct_test_mode/src/dtm.c

Reply:
I use GPPI where possible, but channel allocation does not seem to be covered by GPPI for some reason. 
I discussed this with Krzysztof C, and he confirmed. 

The DTM example uses a hard coded PPI channel, so channel allocation is not used.